python/dgl/data/gdelt.py

```diff
@@ -1,97 +1,167 @@
-                self.dir, 'GDELT', 'valid.txt'), delimiter='\t').astype(np.int64)
-            test_data = loadtxt(os.path.join(
-                self.dir, 'GDELT', 'test.txt'), delimiter='\t').astype(np.int64)
-            train_data[:, 3] = -1
-            val_data[:, 3] = -1
-            self._load(np.concatenate(
-                [train_data, val_data, test_data], axis=0))
-
-    def _download(self):
-        for dname in self._url:
-            dpath = os.path.join(
-                self.dir, 'GDELT', self._url[dname.lower()].split('/')[-1])
-            download(self._url[dname.lower()], path=dpath)
-
-    @retry_method_with_fix(_download)
-    def _load(self, data):
+        self.num_nodes = 23033
+        _url = _get_dgl_url('dataset/gdelt.zip')
+        super(GDELTDataset, self).__init__(name='GDELT',
+                                           url=_url,
+                                           raw_dir=raw_dir,
+                                           force_reload=force_reload,
+                                           verbose=verbose)
+
+    def process(self):
+        file_path = os.path.join(self.raw_path, self.mode + '.txt')
+        self.data = loadtxt(file_path, delimiter='\t').astype(np.int64)
+
         # The source code is not released, but the paper indicates there're
         # totally 137 samples. The cutoff below has exactly 137 samples.
-        self.data = data
-        self.time_index = np.floor(data[:, 3]/15).astype(np.int64)
-        self.start_time = self.time_index[self.time_index != -1].min()
-        self.end_time = self.time_index.max()
+        self.time_index = np.floor(self.data[:, 3] / 15).astype(np.int64)
+        self._start_time = self.time_index.min()
+        self._end_time = self.time_index.max()
+
+    def has_cache(self):
+        info_path = os.path.join(self.save_path, self.mode + '_info.pkl')
+        return os.path.exists(info_path)
+
+    def save(self):
+        info_path = os.path.join(self.save_path, self.mode + '_info.pkl')
+        save_info(info_path, {'data': self.data,
+                              'time_index': self.time_index,
+                              'start_time': self.start_time,
+                              'end_time': self.end_time})
+
+    def load(self):
+        info_path = os.path.join(self.save_path, self.mode + '_info.pkl')
+        info = load_info(info_path)
+        self.data, self.time_index, self._start_time, self._end_time = \
+            info['data'], info['time_index'], info['start_time'], info['end_time']
+
+    @property
+    def start_time(self):
+        r""" Start time of events in the temporal graph
+
+        Returns
+        -------
+        int
+        """
+        return self._start_time
+
+    @property
+    def end_time(self):
+        r""" End time of events in the temporal graph
+
+        Returns
+        -------
+        int
+        """
+        return self._end_time
 
     def __getitem__(self, idx):
+        r""" Get graph by index
```

Comment:
Can we give a more detailed info of how idx is calculated? Especially, i = idx + self.start_time